// @paradise-cloud/easy-debrid/dist/index.mjs
// src/client.ts
import axios from "axios";

// src/errors.ts
import { isAxiosError } from "axios";
var EasyDebridError = class extends Error {
  constructor(message) {
    super(message);
  }
};
var toEasyDebridError = (error) => {
  if (isAxiosError(error)) {
    throw new EasyDebridError(
      error.response?.data.message || error.message
    );
  }
  if (error instanceof Error) {
    throw new EasyDebridError(error.message);
  }
  throw new EasyDebridError("An unknown error occurred");
};

// src/utils.ts
var PROD_HOST = "https://easydebrid.com";
var DEV_HOST = "https://dev.easydebrid.com";
function getHost(env) {
  return env === "production" ? PROD_HOST : DEV_HOST;
}
function getApiUrl(env) {
  return `${getHost(env)}/api/v1`;
}

// src/client.ts
var EasyDebridClient = class {
  constructor(options) {
    this.options = options;
    options.env = options.env || "production";
    const baseUrl = getApiUrl(options.env);
    this.apiClient = axios.create({
      baseURL: baseUrl,
      headers: {
        "Content-Type": "application/json",
        authorization: `Bearer ${this.options.accessToken}`
      }
    });
  }
  async getAccountInfo() {
    try {
      const { data } = await this.apiClient.get(
        "/user/details"
      );
      const paidUntil = data?.paid_until ? new Date(data.paid_until * 1e3) : null;
      const transformedData = {
        ...data,
        paid_until: paidUntil
      };
      return transformedData;
    } catch (error) {
      return toEasyDebridError(error);
    }
  }
  async submitCoupon(coupon) {
    try {
      const { data } = await this.apiClient.post(
        "/coupon/submit",
        { coupon }
      );
      return data;
    } catch (error) {
      return toEasyDebridError(error);
    }
  }
  async linkLookup(urls) {
    try {
      const { data } = await this.apiClient.post(
        "/link/lookup",
        { urls }
      );
      return data;
    } catch (error) {
      return toEasyDebridError(error);
    }
  }
  async linkLookupDetails(urls) {
    try {
      const { data } = await this.apiClient.post(
        "/link/lookupdetails",
        { urls }
      );
      return data;
    } catch (error) {
      return toEasyDebridError(error);
    }
  }
  async generateDebridLink(url) {
    try {
      const { data } = await this.apiClient.post(
        "/link/generate",
        { url }
      );
      return data;
    } catch (error) {
      return toEasyDebridError(error);
    }
  }
};

// src/oauth2.ts
import axios2 from "axios";
var VALID_SCOPES = ["full"];
var EasyDebridOauth2Client = class {
  constructor(options) {
    this.options = options;
    options.env = options.env || "production";
    this.serviceUrl = getHost(options.env);
    this.apiClient = axios2.create({
      baseURL: this.serviceUrl,
      headers: {
        "Content-Type": "application/json"
      }
    });
  }
  generateAuthorizationUrl(redirectUri, codeChallenge, codeChallengeMethod, state, scopes) {
    const params = new URLSearchParams();
    params.append("response_type", "code");
    params.append("client_id", this.options.clientId);
    params.append("redirect_uri", redirectUri);
    params.append("code_challenge", codeChallenge);
    params.append("code_challenge_method", codeChallengeMethod);
    if (state) {
      params.append("state", state);
    }
    if (scopes) {
      params.append("scope", scopes.join(" "));
    }
    return `${this.serviceUrl}/oauth/authorize?${params.toString()}`;
  }
  async exchangeCodeForToken(code, redirectUri, codeVerifier) {
    const params = new URLSearchParams();
    params.append("grant_type", "authorization_code");
    params.append("code", code);
    params.append("redirect_uri", redirectUri);
    params.append("code_verifier", codeVerifier);
    if (this.options.clientSecret) {
      params.append("client_secret", this.options.clientSecret);
    }
    try {
      const { data } = await this.apiClient.post(
        "/oauth/token",
        params,
        {
          headers: {
            "Content-Type": "application/x-www-form-urlencoded"
          }
        }
      );
      return data;
    } catch (error) {
      return toEasyDebridError(error);
    }
  }
};

// src/merchant.ts
import axios3 from "axios";
var EasyDebridMerchant = class {
  constructor(options) {
    this.options = options;
    options.env = options.env || "production";
    const baseUrl = getApiUrl(options.env);
    this.apiClient = axios3.create({
      baseURL: baseUrl,
      headers: {
        "Content-Type": "application/json",
        authorization: `Bearer ${this.options.apiKey}`
      }
    });
  }
  async getResellerPrices() {
    try {
      const { data } = await this.apiClient.get(
        "/reseller/prices"
      );
      return data;
    } catch (error) {
      return toEasyDebridError(error);
    }
  }
  async topupResellerBalance(amount) {
    try {
      const { data } = await this.apiClient.post(
        "/reseller/topup",
        { amount }
      );
      return data;
    } catch (error) {
      return toEasyDebridError(error);
    }
  }
  async getResellerBalance() {
    try {
      const { data } = await this.apiClient.get(
        "/reseller/balance"
      );
      return data;
    } catch (error) {
      return toEasyDebridError(error);
    }
  }
  async generateResellerCoupon(days) {
    try {
      const { data } = await this.apiClient.post(
        "/reseller/generatecoupon",
        { days }
      );
      return data;
    } catch (error) {
      return toEasyDebridError(error);
    }
  }
  async listResellerCoupons() {
    try {
      const { data } = await this.apiClient.get(
        "/reseller/coupons"
      );
      const transformedData = {
        coupons: data.coupons.map((coupon) => ({
          ...coupon,
          created: new Date(coupon.created),
          expires: new Date(coupon.expires)
        }))
      };
      return transformedData;
    } catch (error) {
      return toEasyDebridError(error);
    }
  }
};
export {
  EasyDebridClient,
  EasyDebridMerchant,
  EasyDebridOauth2Client,
  VALID_SCOPES
};
